Source/WebCore/accessibility/AXTableHelpers.cpp

```diff
@@ -76,6 +76,24 @@ bool isTableRole(AccessibilityRole role)
     }
 }
 
+bool isARIAGridRow(Element& element)
+{
+    return hasRole(element, "row"_s);
+}
+
+bool isTableRowElement(Element& element)
+{
+    bool isAnonymous = false;
+    CheckedPtr renderer = element.renderer();
+#if USE(ATSPI)
+    isAnonymous = renderer && renderer->isAnonymous();
+#endif
+
+    bool isNonAnonymousRenderTable = is<RenderTableRow>(renderer.get()) && !isAnonymous;
+    bool ariaRoleIsValidForRow = hasRole(element, "row"_s) || hasRole(element, nullAtom());
+    return (isNonAnonymousRenderTable || is<HTMLTableRowElement>(element) || AXTableHelpers::isARIAGridRow(element)) && ariaRoleIsValidForRow;
```

Comment:
I think this would both be more efficient and more clear if we just early-returned like this, since we require `ariaRoleIsValidForRow` for all the other conditions and thus shouldn't compute them first.

`isNonAnonymousRenderTable` is a bit of confusing name when we're checking for a RenderTableRow and not a RenderTable.

```cpp
if (hasRowRole(element))
    return true;

if (!hasRole(element, nullAtom()) {
    // This has a non-row role, so it shouldn't be considered a row.
    return false;
}

if (is<RenderTableRow>(renderer.get()) && !isAnonymous)
    return true;

return is<HTMLTableRowElement>(element));
```